keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,752 @@
+deployment and are not counted in its status, this can cause problems with resource usage and
+scheduling:
+
+
+1. Unnecessary autoscaling of nodes in tight environments and driving up cloud costs. This can hurt
+   especially if multiple deployments are rolled out at the same time, or if a large
+   `.spec.terminationGracePeriodSeconds` value is requested. See the following issues for more
+   details: [kubernetes/kubernetes#95498](https://github.com/kubernetes/kubernetes/issues/95498),
+   [kubernetes/kubernetes#99513](https://github.com/kubernetes/kubernetes/issues/99513),
+   [kubernetes/kubernetes#41596](https://github.com/kubernetes/kubernetes/issues/41596),
+   [kubernetes/kubernetes#97227](https://github.com/kubernetes/kubernetes/issues/97227).
+2. A problem also arises in contentious environments where pods are fighting over resources. This
+   can bring up exponential backoff for not yet started pods into big numbers and unnecessarily
+   delay start of such pods until they pop from the queue when there are computing resources to run
+   them. This can slow down the deployment considerably. This is described in issue
+   [kubernetes/kubernetes#98656](https://github.com/kubernetes/kubernetes/issues/98656). In that
+   issue, the resources were limited by a quota, but this can be due to other reasons as well. This
+   can occur also in high availability scenarios where pods are expected to run only on certain
+   nodes, and pod anti-affinity forbids to run two pods on the same node.
+3. Terminating pods can still do useful work or hold old connections. Users would like to track
+   this work through the deployment's status. See
+   [kubernetes/kubernetes#110171](https://github.com/kubernetes/kubernetes/issues/110171) for more
+   details.
+
+[kubernetes/kubernetes#107920](https://github.com/kubernetes/kubernetes/issues/107920) issue is covering this as well.
+
+### Goals
+
+- Deployments should allow an option to either wait for its pods to terminate before creating new
+  pods, or to create the pods immediately. This should take into consideration the Deployment
+  strategy.
+- Deployments and ReplicaSets should indicate a number of managed terminating pods in their status
+  field.
+
+### Non-Goals
+
+## Proposal
+
+This KEP proposes to introduce a new  `.spec.podReplacementPolicy` field (similar to Job's
+`.spec.podReplacementPolicy` in [kubernetes/enhancements#3939](https://github.com/kubernetes/enhancements/issues/3939))
+that would control how many pods should be present at any given time.
+
+The termination of a Deployment/ReplicaSet pod is always triggered by a pod deletion due to
+an enforced pod field `restartPolicy: Always`.
+
+We are distinguishing between terminating and terminated pods. 
+- Terminating pods are running pods with a `deletionTimestamp`. 
+- Terminated pods are pods with a `deletionTimestamp` that have reached the `Succeeded` or `Failed` phase
+  and are subsequently removed from etcd.
```

Comment:
If the pod is removed from etcd, does that mean that we no longer track them?

Reply:
We will stop tracking them once they reach the terminal phase. There is a small delay before the pod gets removed from etcd though. Not sure if it is important to mention, but some people might find it weird, that the Pod count is higher in this situation (although not running ones!).